datafusion/core/src/execution/memory_manager/mod.rs

```diff
@@ -97,339 +89,154 @@ impl MemoryManagerConfig {
-    /// memory and it is safe to grow without spilling.
-    fn grow(&self, required: usize) {
-        self.memory_manager().record_free_then_acquire(0, required);
-    }
-
-    /// Return `freed` memory to the memory manager,
-    /// may wake up other requesters waiting for their minimum memory quota.
-    fn shrink(&self, freed: usize) {
-        self.memory_manager().record_free(freed);
-    }
-
-    /// Spill in-memory buffers to disk, free memory, return the previous used
-    async fn spill(&self) -> Result<usize>;
-
-    /// Current memory used by this consumer
-    fn mem_used(&self) -> usize;
 }
 
-impl Debug for dyn MemoryConsumer {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        write!(
-            f,
-            "{}[{}]: {}",
-            self.name(),
-            self.id(),
-            human_readable_size(self.mem_used())
-        )
-    }
-}
-
-impl Display for dyn MemoryConsumer {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        write!(f, "{}[{}]", self.name(), self.id(),)
-    }
-}
-
-/*
-The memory management architecture is the following:
-
-1. User designates max execution memory by setting RuntimeConfig.max_memory and RuntimeConfig.memory_fraction (float64 between 0..1).
-   The actual max memory DataFusion could use `pool_size =  max_memory * memory_fraction`.
-2. The entities that take up memory during its execution are called 'Memory Consumers'. Operators or others are encouraged to
-   register themselves to the memory manager and report its usage through `mem_used()`.
-3. There are two kinds of consumers:
-   - 'Requesting' consumers that would acquire memory during its execution and release memory through `spill` if no more memory is available.
-   - 'Tracking' consumers that exist for reporting purposes to provide a more accurate memory usage estimation for memory consumers.
-4. Requesting and tracking consumers share the pool. Each controlling consumer could acquire a maximum of
-   (pool_size - all_tracking_used) / active_num_controlling_consumers.
-
-            Memory Space for the DataFusion Lib / Process of `pool_size`
-   ┌──────────────────────────────────────────────z─────────────────────────────┐
-   │                                              z                             │
-   │                                              z                             │
-   │               Requesting                     z          Tracking           │
-   │            Memory Consumers                  z       Memory Consumers      │
-   │                                              z                             │
-   │                                              z                             │
-   └──────────────────────────────────────────────z─────────────────────────────┘
-*/
-
-/// Manage memory usage during physical plan execution
+/// The memory manager maintains a fixed size pool of memory
+/// from which portions can be allocated
```

Comment:
I think it would be good to expand these comments to explain the allocation strategy of the manager

```suggestion
/// from which portions can be allocated
///
/// A cooperative MemoryManager which tracks memory in a cooperative fashion. 
/// `ExectionPlan` nodes such as `SortExec`, which require large amounts of memory register their
/// memory requests with the MemoryManager which then tracks the total memory that 
/// has been allocated across all such nodes. 
///
/// The MemoryManager does not guarantee any form of "fairness" -- the operators that ask for 
/// memory first are the ones that are given memory. 
///
/// Note that more sophisticated memory polices can be created at a higher level
///  by dividing available system memory across multiple `MemoryManagers`  and assigning 
/// different MemoryManagers to diffrent pools. 
```